File: finance_core/display.py
from __future__ import annotations

import unicodedata


def display_width(s: str) -> int:
    return sum(2 if unicodedata.east_asian_width(c) in ("W", "F") else 1 for c in s)


def fit(s: str, cols: int) -> str:
    """cols 表示列幅に収まるよう切り詰め、右側をスペースで埋める。"""
    result, width = [], 0
    for c in s:
        cw = 2 if unicodedata.east_asian_width(c) in ("W", "F") else 1
        if width + cw > cols:
            break
        result.append(c)
        width += cw
    return "".join(result) + " " * (cols - width)

File: cfor_rs/src/display.rs
use unicode_width::UnicodeWidthChar;

pub fn display_width(s: &str) -> usize {
    s.chars()
        .map(|c| UnicodeWidthChar::width(c).unwrap_or(1))
        .sum()
}

pub fn fit(s: &str, cols: usize) -> String {
    let mut result = String::new();
    let mut width = 0usize;
    for c in s.chars() {
        let cw = UnicodeWidthChar::width(c).unwrap_or(1);
        if width + cw > cols {
            break;
        }
        result.push(c);
        width += cw;
    }
    let padding = cols.saturating_sub(width);
    result.push_str(&" ".repeat(padding));
    result
}

pub fn right_align_amount(amount: i64, width: usize) -> String {
    let s = format!("{}円", crate::models::format_amount(amount));
    let w = display_width(&s);
    if w >= width {
        s
    } else {
        format!("{}{}", " ".repeat(width - w), s)
    }
}
